### PR TITLE
Wip/altran/default shortcuts

### PR DESCRIPTION
**Pull request by altran01**
Add default shortcuts to existing actions and add some new actions (close tab, switch next/prev tab) with default shortcuts.

### CODE REVIEW

**Comment by mkow** [comment]
Could you also bind:
* Ctrl+G to "jump to address"
* More shortcuts to 3d view (e.g. switching between visualization modes)

File: src/ui/dockwidget.cc [hunk 1]
```diff
@@ -111,6 +114,22 @@ DockWidget::DockWidget() : QDockWidget(), timer_id_(0), ticks_(0),
         this, SLOT(displayContextMenu(const QPoint&)));
   connect(this, SIGNAL(topLevelChanged(bool)),
       this, SLOT(topLevelChangedNotify(bool)), Qt::QueuedConnection);
+
+  next_tab_action_ = ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::SWITCH_TAB_NEXT, this,
+        Qt::WidgetWithChildrenShortcut);
+  connect(next_tab_action_, &QAction::triggered, [this]() {
+    MainWindowWithDetachableDockWidgets::focusNextPrevDock(this, true);
```
Comment (mkow) [inline src/ui/dockwidget.cc]: please comment meaning of the bool argument

File: src/ui/dockwidget.cc [hunk 2]
```diff
@@ -111,6 +114,22 @@ DockWidget::DockWidget() : QDockWidget(), timer_id_(0), ticks_(0),
...
+  connect(next_tab_action_, &QAction::triggered, [this]() {
+    MainWindowWithDetachableDockWidgets::focusNextPrevDock(this, true);
+  });
+  addAction(next_tab_action_);
+
+  prev_tab_action_ = ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::SWITCH_TAB_PREV, this,
+        Qt::WidgetWithChildrenShortcut);
+  connect(prev_tab_action_, &QAction::triggered, [this]() {
+    MainWindowWithDetachableDockWidgets::focusNextPrevDock(this, false);
```
Comment (mkow) [inline src/ui/dockwidget.cc]: ditto

File: src/ui/dockwidget.cc [hunk 3]
```diff
@@ -662,6 +704,46 @@ void View::deleteIcons() {
   icons_.clear();
 }
 
+void View::createVisualization(MainWindowWithDetachableDockWidgets* main_window,
+                          QSharedPointer<FileBlobModel> data_model) {
```
Comment (mkow) [inline src/ui/dockwidget.cc]: fix indent

File: src/ui/dockwidget.cc [hunk 4]
```diff
@@ -662,6 +704,46 @@ void View::deleteIcons() {
   icons_.clear();
 }
 
+void View::createVisualization(MainWindowWithDetachableDockWidgets* main_window,
+                          QSharedPointer<FileBlobModel> data_model) {
+  auto *panel = new visualization::VisualizationPanel(main_window, data_model);
+  panel->setData(QByteArray((const char *)data_model->binData().rawData(),
+    static_cast<int>(data_model->binData().size())));
```
Comment (mkow) [inline src/ui/dockwidget.cc]: indent with 4 spaces instead of 2

Comment (mkow) [inline src/ui/dockwidget.cc]: use static_cast in both places

File: src/ui/dockwidget.cc [hunk 5]
```diff
@@ -662,6 +704,46 @@ void View::deleteIcons() {
...
+//  if (sibling == nullptr) {
+//    main_window->addDockWidget(Qt::RightDockWidgetArea, dock_widget);
+//  }
+}
+
+void View::createHexEditor(MainWindowWithDetachableDockWidgets* main_window,
+                      QSharedPointer<FileBlobModel> data_model) {
+  QSharedPointer<QItemSelectionModel> new_selection_model(
+        new QItemSelectionModel(data_model.data()));
+  NodeWidget *node_edit = new NodeWidget(main_window, data_model,
```
Comment (mkow) [inline src/ui/dockwidget.cc]: `*` to the left

File: src/ui/dockwidget.cc [hunk 6]
```diff
@@ -662,6 +704,46 @@ void View::deleteIcons() {
...
+      new_selection_model);
+
+  // FIXME: main_window_ needs to be updated when docks are moved around,
+  // then we can use this behaviour without any weird effects
+  // auto sibling = DockWidget::getParentDockWidget(this);
+  DockWidget* sibling = nullptr;
+
+  auto dock_widget = main_window->addTab(
+        node_edit, data_model->path().join(" : "), sibling);
```
Comment (mkow) [inline src/ui/dockwidget.cc]: indent with 4 spaces

File: src/ui/dockwidget.cc [hunk 7]
```diff
@@ -1051,6 +1134,23 @@ void MainWindowWithDetachableDockWidgets::setActiveDockWidget(
...
+  auto parent = getOwnerOfDockWidget(dock_widget);
+
+  if (parent) {
+    auto tab_pair = parent->dockWidgetToTab(dock_widget);
+    if (tab_pair.first) {
+      int index;
+      if (next) {
+        index = (tab_pair.second + 1) % tab_pair.first->count();
+      } else {
+        index = (tab_pair.second - 1 + tab_pair.first->count())  % tab_pair.first->count();
```
Comment (mkow) [inline src/ui/dockwidget.cc]: fix double space

File: src/util/settings/shortcuts.cc [hunk 8]
```diff
@@ -365,6 +372,9 @@ ShortcutsModel::ShortcutsModel() : root_(new ShortcutsItem()) {
   addShortcutType(DOCK_MOVE_TO_TOP_MAX, docks, tr("Move to new top level window and maximize"));
   addShortcutType(DOCK_SPLIT_HORIZ, docks, tr("Split horizontally"));
   addShortcutType(DOCK_SPLIT_VERT, docks, tr("Split vertically"));
+  addShortcutType(DOCK_CLOSE, docks, tr("Close current dock"));
+  addShortcutType(SWITCH_TAB_NEXT, docks, tr("Switch to the next Tab"));
```
Comment (mkow) [inline src/util/settings/shortcuts.cc]: why `Tab` instead of `tab`?

File: src/util/settings/shortcuts.cc [hunk 9]
```diff
@@ -365,6 +372,9 @@ ShortcutsModel::ShortcutsModel() : root_(new ShortcutsItem()) {
...
+  addShortcutType(SWITCH_TAB_PREV, docks, tr("Switch to the previous Tab"));
```
Comment (mkow) [inline src/util/settings/shortcuts.cc]: ditto

File: src/visualization/panel.cc [hunk 10]
```diff
@@ -387,6 +387,27 @@ void VisualizationPanel::initOptionsPanel() {
   /////////////////////////////////////
   // Additional toolbars specific for current visualization
   prepareVisualizationOptions();
+
+  // TODO considering that there is already some code duplication between here
+  // and HexEditWidget we should consider refactoring some parts of those classes
+  // into common base
+  QAction* open_hex = ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::OPEN_HEX, this,
```
Comment (mkow) [inline src/visualization/panel.cc]: fix indent

File: src/visualization/panel.cc [hunk 11]
```diff
@@ -387,6 +387,27 @@ void VisualizationPanel::initOptionsPanel() {
...
+        Qt::WidgetWithChildrenShortcut);
+
+  connect(open_hex, &QAction::triggered, [this]() {
+    createHexEditor(main_window_, data_model_);
+  });
+  addAction(open_hex);
+
+  QAction* open_visualization = ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::OPEN_VISUALIZATION, this,
```
Comment (mkow) [inline src/visualization/panel.cc]: fix indent

File: src/visualization/panel.cc [hunk 12]
```diff
@@ -387,6 +387,27 @@ void VisualizationPanel::initOptionsPanel() {
...
+
+  // TODO considering that there is already some code duplication between here
+  // and HexEditWidget we should consider refactoring some parts of those classes
+  // into common base
+  QAction* open_hex = ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::OPEN_HEX, this,
+        Qt::WidgetWithChildrenShortcut);
+
+  connect(open_hex, &QAction::triggered, [this]() {
+    createHexEditor(main_window_, data_model_);
```
Comment (mkow) [inline src/visualization/panel.cc]: fix indent

File: src/visualization/panel.cc [hunk 13]
```diff
@@ -387,6 +387,27 @@ void VisualizationPanel::initOptionsPanel() {
...
+  });
+  addAction(open_hex);
+
+  QAction* open_visualization = ShortcutsModel::getShortcutsModel()->createQAction(
+        util::settings::shortcuts::OPEN_VISUALIZATION, this,
+        Qt::WidgetWithChildrenShortcut);
+
+  connect(open_visualization, &QAction::triggered, [this]() {
+    createVisualization(main_window_, data_model_);
```
Comment (mkow) [inline src/visualization/panel.cc]: fix indent

File: src/ui/dockwidget.cc (hunk 4 above)
Comment (altran01) [inline src/ui/dockwidget.cc]: Both were already present in code that I moved here, but w/e :P

File: src/ui/dockwidget.cc (hunk 5 above)
Comment (altran01) [inline src/ui/dockwidget.cc]: ditto.

File: src/ui/dockwidget.cc [hunk 14]
```diff
@@ -119,15 +119,15 @@ DockWidget::DockWidget() : QDockWidget(), timer_id_(0), ticks_(0),
         util::settings::shortcuts::SWITCH_TAB_NEXT, this,
         Qt::WidgetWithChildrenShortcut);
   connect(next_tab_action_, &QAction::triggered, [this]() {
-    MainWindowWithDetachableDockWidgets::focusNextPrevDock(this, true);
+    MainWindowWithDetachableDockWidgets::focusNextPrevDock(this, /*next*/true);
```
Comment (mkow) [inline src/ui/dockwidget.cc]: `/*next=*/`

File: src/ui/dockwidget.cc [hunk 15]
```diff
@@ -119,15 +119,15 @@ DockWidget::DockWidget() : QDockWidget(), timer_id_(0), ticks_(0),
...
   });
   addAction(next_tab_action_);
 
   prev_tab_action_ = ShortcutsModel::getShortcutsModel()->createQAction(
         util::settings::shortcuts::SWITCH_TAB_PREV, this,
         Qt::WidgetWithChildrenShortcut);
   connect(prev_tab_action_, &QAction::triggered, [this]() {
-    MainWindowWithDetachableDockWidgets::focusNextPrevDock(this, false);
+    MainWindowWithDetachableDockWidgets::focusNextPrevDock(this, /*next*/false);
```
Comment (mkow) [inline src/ui/dockwidget.cc]: ditto

**Pull request merged**